File: flask_app/app/services/auth_service.py
from app.utils.security import hash_password, verify_password
from app.dao.auth_dao import AuthDAO
from app import db
from datetime import datetime


class AuthService:

    @staticmethod
    def register(name, email, password, phone_number=None, date_of_birth=None):
        existing_user = AuthDAO.get_by_email(email)
        if existing_user:
            raise Exception("User already exists")

        hashed_password = hash_password(password)
        dob = None
        if date_of_birth:
            try:
                dob = datetime.strptime(date_of_birth, "%Y/%m/%d")  # or "%Y-%m-%d"
            except ValueError:
                raise Exception("Invalid date format. Use YYYY/MM/DD.")
            
        return AuthDAO.create_user(
            name=name,
            email=email,
            password_hash=hashed_password,
            phone_number=phone_number,
            date_of_birth=dob
        )

    @staticmethod
    def login(email, password):
        user = AuthDAO.get_by_email(email)
        if user and verify_password(user.password_hash, password):
            return user
        return None

    @staticmethod
    def update_last_login(user_id):
        user = AuthDAO.get_by_id(user_id)
        if user:
            user.last_login = datetime.utcnow()
            db.session.commit()

    @staticmethod
    def get_user_by_id(user_id):
        return AuthDAO.get_by_id(user_id)

    @staticmethod
    def change_password(user_id, current_password, new_password):
        user = AuthDAO.get_by_id(user_id)
        if not user:
            return None, "User not found"

        if not verify_password(user.password_hash, current_password):
            return None, "Incorrect current password"

        updated_user = AuthDAO.change_password(user_id, new_password)
        return updated_user, None

File: java_membership_api/src/main/java/com/example/membership_api/service/AuthService.java
package com.example.membership_api.service;

import java.time.LocalDateTime;

import org.apache.commons.lang3.tuple.Pair;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import com.example.membership_api.controller.HealthController;
import com.example.membership_api.dao.AuthDAO;
import com.example.membership_api.model.User;
import com.example.membership_api.utils.PasswordUtils;




@Service
public class AuthService {

    private static final Logger logger = LoggerFactory.getLogger(HealthController.class);

    @Autowired
    private AuthDAO authDAO;

    public User register(String name, String email, String password, String phone, LocalDateTime dob) {
        if (authDAO.getByEmail(email) != null) {
            throw new IllegalArgumentException("User already exists");
        }

        logger.info( "Service: name {}, email {}, phone {}, dob {}, password {} Hash {}", name, email, phone, dob, password, PasswordUtils.hashPassword(password) );

        User user = new User();
        user.setName(name);
        user.setEmail(email);
        user.setPasswordHash(PasswordUtils.hashPassword(password));
        user.setPhoneNumber(phone);
        user.setDateOfBirth(dob);
        user.setCreatedAt(LocalDateTime.now());
        logger.info("user {}", user.toString());
        return authDAO.createUser(user);
    }

    public User login(String email, String password) {
        User user = authDAO.getByEmail(email);
        if (user != null && PasswordUtils.verifyPassword(password, user.getPasswordHash())) {
            return user;
        }
        return null;
    }

    public void updateLastLogin(String userId) {
        User user = authDAO.getById(userId);
        if (user != null) {
            user.setLastLogin(LocalDateTime.now());
        }
    }

    public User getUserById(String id) {
        return authDAO.getById(id);
    }

    public Pair<User, String> changePassword(String id, String currentPassword, String newPassword) {
        User user = authDAO.getById(id);
        if (user == null) return Pair.of(null, "User not found");

        if (!PasswordUtils.verifyPassword(currentPassword, user.getPasswordHash())) {
            return Pair.of(null, "Incorrect current password");
        }

        User updated = authDAO.changePassword(id, PasswordUtils.hashPassword(newPassword));
        return Pair.of(updated, null);
    }
}
